### PR TITLE
GitHub: easily add upstream remote

### PR DESCRIPTION
**Pull request by pmiossec**
## Proposed changes

~~Note: better reviewable and merged after #6514 (a small part of the refactoring done in this PR was extracted in #6514)~~ done.

- Add a menu item to easily add the "upstream" remote to the repository
- refactoring:  also introduce a refactoring on `GitHubHostedRemote` where the underneath repository should be initialized with the `GitProtocol` already used ( make use of `GitHubHostedRemote` easier)

## Screenshots 

image

## Test methodology 

- Manual

## Test environment(s) 

- Git Extensions 3.1.0
- Build 9c567e63a5bfb30af412bce86e356e169735d5af
- Git 2.20.0.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3324.0
- DPI 192dpi (200% scaling)

### CODE REVIEW

File: Plugins/GitHub3/GitHubRepo.cs [hunk 1]
```diff
@@ -131,7 +131,10 @@ public int CreatePullRequest(string myBranch, string remoteBranch, string title,
 
         public string CloneProtocol { get; set; } = "SSH";
 
-        public string[] SupportedCloneProtocols { get; set; } = new string[] { "SSH", "HTTPS" };
+        public const string SshProtocol = "SSH";
+        public const string HttpsProtocol = "HTTPS";
+        public string[] SupportedCloneProtocols { get; set; } = { SshProtocol, HttpsProtocol };
```
Comment (RussKie) [inline Plugins/GitHub3/GitHubRepo.cs]: Retuning arrays is rather dangerous, let's fix it
```suggestion
        public IReadOnlyList<string> SupportedCloneProtocols { get; set; } = { SshProtocol, HttpsProtocol };
```

File: Plugins/GitHub3/GitHubRepo.cs [hunk 2]
```diff
@@ -131,7 +131,10 @@ public int CreatePullRequest(string myBranch, string remoteBranch, string title,
...
+        public string ParentCloneReadWriteUrl => CloneProtocol == "SSH" ? _repo.Parent?.SshUrl : _repo.Parent?.CloneUrl;
```
Comment (RussKie) [inline Plugins/GitHub3/GitHubRepo.cs]: Why not use the const created above?

**Comment by pmiossec** [comment]
Done. I have also added the fetch after the remote added.

File: Plugins/GitHub3/GitHub3Plugin.cs [hunk 3]
```diff
@@ -145,6 +148,31 @@ public IReadOnlyList<IHostedRepository> GetMyRepos()
             return GitHub.getRepositories().Select(repo => (IHostedRepository)new GitHubRepo(repo)).ToList();
         }
 
+        public string OwnerLogin => GitHub.getCurrentUser()?.Login;
+
+        public async Task<string> AddUpstreamRemoteAsync(IGitModule module)
```
Comment (RussKie) [inline Plugins/GitHub3/GitHub3Plugin.cs]: ⚠️  We mustn't do it like this.

A git module is passed via `Register` event:
https://github.com/gitextensions/gitextensions/blob/98747b18124c482187b4a0f06b94df985103e12c/Plugins/GitHub3/GitHub3Plugin.cs#L101-L107

**Comment by RussKie** [comment]
Apologies for changing my mind, I have missed that previously

File: Plugins/GitHub3/GitHub3Plugin.cs (hunk 3 above)
Comment (pmiossec) [reply Plugins/GitHub3/GitHub3Plugin.cs]: Done. But that obliged me to do the same thing for the other already existing methods (in a new commit) 😉

**Comment by RussKie** [comment]
I pushed an update to translations.
You may need to be rebase, not sure

**Comment by pmiossec** [comment]
Done

**Comment by RussKie** [comment]
Thank you

**Pull request merged**